[TensorExpr] Add IR visitor, IR mutator, and IR evaluator.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33220 [TensorExpr] Add IR Printer.
* **#33219 [TensorExpr] Add IR visitor, IR mutator, and IR evaluator.**
* #33218 [TensorExpr] Add core classes for representing expressions and statements.
* #33217 [TensorExpr] Add a class for representing data type.
* #33216 [TensorExpr] Add classes for memory management in tensor expressions.



Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

Commit:
[TensorExpr] Add IR visitor, IR mutator, and IR evaluator.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Stamping as it's pretty straightforward and has been validated by additional code built on top in https://github.com/bertmaher/pytorch

Commit:
Rebase on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."


Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."




Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]

Commit:
Rebase. on "[TensorExpr] Add IR visitor, IR mutator, and IR evaluator."




Differential Revision: [D19848381](https://our.internmc.facebook.com/intern/diff/D19848381)

[ghstack-poisoned]
